cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/file/FileSetDataset.java

```diff
@@ -274,8 +274,10 @@ public void onFailure() throws DataSetException {
       // but it leaves around the directory that it created.
       // We don't want to unconditionally delete the output directory on failure, because it may have files written
       // by a different job.
-      if (outputLocation.isDirectory() && !outputLocation.delete()) {
-        throw new DataSetException(String.format("Error deleting file(s) at path %s.", outputLocation));
+      if (outputLocation.isDirectory() && outputLocation.list().isEmpty()) {
+        if (!outputLocation.delete()) {
```

Comment:
if you want to avoid making too making too many calls, you could move the isEmpty() into here, and only throw the exception if it is empty. Perhaps that would address @chtyim 's concern